BJS2-3311 - Организация события - Unicorn BC3

hhzktoeto opened the pull request.

hhzktoeto committed:
BJS2-3311 (NO TESTS) - Организация события
(Создать событие)

Реализовал логику без Unit-tests

hhzktoeto committed:
В EventController в метод validateEventDto добавил ещё одну валидацию к полю title

hhzktoeto committed:
Переименовал метод userHasRequiredSkills в userHasRequiredSkillsValidation

hhzktoeto committed:
Изменил методы для валидации. Повысил читаемость кода

hhzktoeto committed:
BJS2-3311 - Организация события (Создать событие)

1. Закончил работу над подтаской, написал 4 теста:
- SkillMapperTest
- EventMapperTest
- EventControllerTest
- EventServiceTest

2. Добавил аннотацию @Builder на EventDto и SkillDto

3. EventController.create больше не пробрасывает DataValidationException выше

hhzktoeto committed:
Временно

hhzktoeto committed:
added updateEvent method

hhzktoeto committed:
Выполнил 5 подтасок

Реализовал возможности:
- создать событие;
- получить событие;
- получить все события с фильтрами;
- удалить событие;
- обновить событие.
Написал тесты на все методы.

Создал класс EventValidator, валидацию из всего кода вынес в этот класс, написал на него тесты.

Общий рефакторинг и доработки

hhzktoeto committed:
Мелкие поправки

hhzktoeto committed:
Мелкие поправки

hhzktoeto committed:
Закончил работу над задачей

Реализовал возможности:
- получить все созданные пользователем события;
- получить все события, в которых пользователь принимает участие.
Написал тесты на все методы

Добавил UserValidator, написал на него тесты

snailrake commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,86 @@
+    private final EventValidator eventValidator;
+    private final UserValidator userValidator;
+
+    public EventDto create(EventDto eventDto) {
+        eventValidator.validateUserHasRequiredSkills(eventDto);
+        eventRepository.save(eventMapper.toEntity(eventDto));
+        return eventDto;
+    }
```
Метод save возвращает сущность из БД, желательно её смаппить в ДТО и вернуть, а не просто пришедшую ДТО возвращать. К примеру, у сохранённого в БД ивента мог появиться айдишник или что-то другое, что мы тоже хотим вернуть

hhzktoeto replied:
Исправил

snailrake commented on src/main/java/school/faang/user_service/service/event/filter/EventTitleFilter.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public void apply(List<Event> events, EventFilterDto filters) {
+        events.removeIf(event -> event.getTitle().contains(filters.getTitle()));
```
Ремуваем если название не содержит паттерна из фильтра, а не если содержит

hhzktoeto replied:
Исправил, оч хорошо что заметил

snailrake commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,86 @@
+        Event eventToUpdate = eventRepository.findById(eventDto.getId()).get(); //validateEventExistsById выкинет ошибку, если там null
+        eventValidator.validateUserIsOwnerOfEvent(eventToUpdate.getOwner(), eventDto);
+
+        Event updatedEvent = setUpdatedInformation(eventToUpdate, eventDto);
```
Можно сделать метод setUpdatedInformation void и просто сохранять пришедший из БД ивент

hhzktoeto replied:
Исправил

hhzktoeto committed:
Исправления после Cross review

hhzktoeto committed:
Убрал ненужный импорт

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/dto/event/EventDto.java:
```diff
@@ -0,0 +1,26 @@
+    private LocalDateTime endDate;
+    private Long ownerId;
+    private String description;
+    private List<SkillDto> relatedSkills;
```
в дто всегда передаем айдишники сущностей, а не целиком
Дто в целом тяжело тащить, а особенно пользователю будет неудобно заполнять так - поэтому тут просто айди клади, и тебя будет список айдишников

hhzktoeto replied:
Я с тобой абсолютно согласен, только в задаче написано передавать List\<SkillDto\>
![image](https://github.com/CorporationX/user_service/assets/153752474/e5794325-b245-494d-93ed-28a42ea00f28)

И у прошлых потоков тоже в Dto лежит List\<SkillDto\>, вот я и оставил так

IrinaVereshchagina13 replied:
Поняла, давай все же заменим, на List<Long>
Пойдем против системы:))

IrinaVereshchagina13 replied:
И в EventFilterDto соответственно тоже

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/dto/event/EventFilterDto.java:
```diff
@@ -0,0 +1,25 @@
+    private LocalDateTime endDate;
+    private Long ownerId;
+    private String description;
+    private List<SkillDto> relatedSkills;
```
тут тоже айди

hhzktoeto replied:
Ответил выше

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,20 @@
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface SkillMapper {
+
+    SkillDto toDto(Skill skill);
+
+    Skill toEntity(SkillDto skillDto);
+
+    List<SkillDto> toDto(List<Skill> skills);
+
+    List<Skill> toEntity(List<SkillDto> skillsDto);
```
а ты все эти методы у себя используешь?

hhzktoeto replied:
Никак нет, но написал реализацию сразу, если в будущем понадобится. 
Размышлял так: мне две строки кода написать не трудно, но зато я к этому мапперу больше возвращаться не буду. Написал и забыл.

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,85 @@
+
+    public EventDto getEvent(long eventId) {
+        eventValidator.validateEventExistsById(eventId);
+        Event searchedEvent = eventRepository.findById(eventId).get(); //validateEventExistsById выкинет ошибку, если там null
```
через гет не делаем  
пытаешься достать, если нет такой сущность, что через orElseThrow бросаешь подходяшее исключение здесь

hhzktoeto replied:
Я ведь написал отдельный метод для проверки - существует ли такой ивент :(
eventValidator.validateEventExistsById(eventId);
Это позволяет не писать конструкцию orElseThrow, которая на мой взгляд очень сильно засоряет код и позволяет тестировать работу метода отдельно от валидации

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public void deleteEvent(long eventId) {
+        eventValidator.validateEventExistsById(eventId);
+        eventRepository.deleteById(eventId);
+    }
```
delete ничего не удалит из базы, если такой записи там и не было 
ПОэтому валидацию можно не делать здесь

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,85 @@
+    public EventDto updateEvent(EventDto eventDto) {
+        eventValidator.validateUserHasRequiredSkills(eventDto);
+        eventValidator.validateEventExistsById(eventDto.getId());
+        Event event = eventRepository.findById(eventDto.getId()).get(); //validateEventExistsById выкинет ошибку, если там null
```
то же, что выше

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,85 @@
+        eventValidator.validateUserIsOwnerOfEvent(event.getOwner(), eventDto);
+
+        setUpdatedInformation(event, eventDto);
+        Event updatedAndSavedEvent = eventRepository.save(event);
```
ты можешь не сетить новую информацию в отдельном методе 
save обновляет все поля по айди, если такая запись в бд уже была

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,85 @@
+        return eventMapper.toDto(userParticipatedEvents);
+    }
+
+    private void setUpdatedInformation(Event eventToUpdate, EventDto eventDto) {
```
вот это тогда тебе не понадобится

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/validation/event/EventValidator.java:
```diff
@@ -0,0 +1,59 @@
+    private final SkillMapper skillMapper;
+
+    public void validateEventDtoFields(EventDto eventDto) {
+        if (eventDto.getTitle() == null || eventDto.getTitle().isEmpty() || eventDto.getTitle().isBlank()) {
```
посмотри как работает IsEmpty и isBlank - тебе только один из них нужен, оба перебор)

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/validation/event/EventValidator.java:
```diff
@@ -0,0 +1,59 @@
+
+    public void validateUserIsOwnerOfEvent(User user, EventDto eventDto) {
+        if (user.getId() != eventDto.getOwnerId()) {
+            throw new IllegalStateException("User is not an owner of the Event");
```
создай себе DataValidationException и используй для всех валидаций вообще, где будет подходить -чтобы везде одинаково было

hhzktoeto committed:
Исправления после ревью

- Убрал аннотации @SpringBootTest со всех тестов, заменил на @ExtendsWith
- Удалил валидацию на существование ивента по id в классе EventService методе deleteEvent
- Убрал излишнюю проверку на isEmpty в классе EventValidator в методе validateEventDtoFields
- В классе EventService удалил метод setUpdatedInformation
- Теперь все методы валидации выбрасывают одинаковый DataValidationException
- Почистил импорты

Mihusle committed:
Removed wrongly merged test from another branch

Mihusle committed:
Fixed getting a specific user skill

hhzktoeto committed:
Финальные поправки

hhzktoeto committed:
Очистил ненужные импорты
